lib/language_pack/ruby.rb

```diff
@@ -117,6 +118,66 @@ def compile
     raise e
   end
 
+  def patch_libpq
+    # Check for existing libraries
+    out = run!("ls /usr/lib/x86_64-linux-gnu/ | grep libpq")
+    out.each_line do |line|
+      version = line.sub(/libpq\.so\./, '')
+      next unless version.match?(/\Ad/) # avoid libpq.so.libpq.a
+      return if Gem::Version.new(version) >= Gem::Version("5.12")
+    end
+
+    case ENV['STACK']
+    when 'heroku-16'
+      pkg = "libpq5_12.1-1.pgdg16.04+1_amd64.deb"
+    when 'heroku-18'
+      pkg = "libpq5_12.1-1.pgdg18.04+1_amd64.deb"
+    when 'cedar-14'
+      return
+    else
+      return
+    end
+
+    @metadata.touch("patched_libpq12")
+
+    topic("Vendoring libpq 5.12.1")
+    warn(<<~EOF)
+      Replacing libpq with version libpq 5.12.1
+
+      This version includes a bug fix that can cause an exception
+      on boot for applications with incorrectly configured connection
+      values. For more information see:
+
+        https://devcenter.heroku.com/articles/libpq-5-12-1-breaking-connection-behavior
+
+      If your application breaks you can rollback to your last build.
+      You can also temporarially opt out of this behavior by setting:
+
+      ```
+      $ heroku config:set HEROKU_SKIP_LIBPQ12=1
+      ```
+
+      In the future libpq 5.12 will be the default on the platform and
+      you will not be able to opt-out of the library. For more information see:
+
+        https://devcenter.heroku.com/articles/libpq-5-12-1-breaking-connection-behavior
+    EOF
+
+    Dir.chdir("vendor") do
+      run!("curl --remote-name-all http://apt.postgresql.org/pub/repos/apt/pool/main/p/postgresql-12/#{pkg}")
```

Comment:
As discussed on Slack it would be good to download this from S3 rather than a third party domain that may be down during the customer's build. Ideally the S3 upload should be performed by a script in this repo that can be reviewed and so be compliant. In that script I would also like for us to download from `https://apt.postgresql.org` rather than the non-HTTPS URL :-)

Reply:
Yes, we have to mirror to our S3 bucket rather than downloading straight off a third party server.